cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs

```diff
@@ -0,0 +1,170 @@
+
+/// 1. Opening a channel. Step 1 of handshake. Combines ChanOpenInit and ChanOpenTry from the spec.
+/// The only valid action of the contract is to accept the channel or reject it.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_open(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelOpenMsg,
+) -> StdResult<IbcChannelOpenResponse> {
+    let channel = msg.channel();
+
+    if channel.version.as_str() != IBC_APP_VERSION {
+        return Err(StdError::generic_err(format!(
+            "Must set version to `{}`",
+            IBC_APP_VERSION
+        )));
+    }
+
+    if let Some(counter_version) = msg.counterparty_version() {
+        if counter_version != IBC_APP_VERSION {
+            return Err(StdError::generic_err(format!(
+                "Counterparty version must be `{}`",
+                IBC_APP_VERSION
+            )));
+        }
+    }
+
+    // We return the version we need (which could be different than the counterparty version)
+    Ok(Some(Ibc3ChannelOpenResponse {
+        version: IBC_APP_VERSION.to_string(),
+    }))
+}
+
+/// 2. Step 2 of handshake. Combines ChanOpenAck and ChanOpenConfirm from the spec.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_connect(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelConnectMsg,
+) -> StdResult<IbcBasicResponse> {
+    let channel = msg.channel();
+    let connection_id = &channel.connection_id;
+
+    Ok(IbcBasicResponse::new()
+        .add_attribute("action", "ibc_connect")
+        .add_attribute("connection_id", connection_id))
+}
+
+/// 3. Closing a channel - whether due to an IBC error, at our request, or at the request of the other side.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_close(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelCloseMsg,
+) -> StdResult<IbcBasicResponse> {
+    let channel = msg.channel();
+    let channel_id = &channel.endpoint.channel_id;
+
+    Ok(IbcBasicResponse::new()
+        .add_attribute("action", "ibc_close")
+        .add_attribute("channel_id", channel_id))
+}
+
+/// 4. Receiving a packet.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_packet_receive(
+    deps: DepsMut,
+    _env: Env,
+    msg: IbcPacketReceiveMsg,
+) -> StdResult<IbcReceiveResponse> {
+    // decode the packet as WormholeIbcPacketMsg::Publish and add the response attributes to the IbcReceiveResponse
+    // put this in a closure so we can convert all error responses into acknowledgements
+    (|| {
+        let packet = msg.packet;
+        // which local channel did this packet come on
+        let channel_id = packet.dest.channel_id;
+        let wormhole_msg: WormholeIbcPacketMsg = from_slice(&packet.data)?;
+        match wormhole_msg {
+            WormholeIbcPacketMsg::Publish { msg: publish_msg } => {
+                receive_publish(deps, channel_id, publish_msg)
+            }
+        }
+    })()
```

Comment:
style nit: I'd prefer if we move this out into a named function (like `handle_ibc_packet`) rather than inlining it as a closure here.

Reply:
can do. Any particular reason why you feel that would be better style?